foxpuppet/windows/browser/panel_ui/panel_ui.py

```diff
@@ -0,0 +1,288 @@
+
+        Returns:
+            list: A list of links that were not found in the awesome bar.
+        """
+        if new_window:
+            self.open_new_window()
+        else:
+            self.open_new_tab()
+        self.switch_to_new_window_or_tab()
+        for link in links:
+            self.selenium.get(link)
+        return self.awesome_bar(links)
+
+    def verify_private_browsing_links_not_in_awesome_bar(self, links: list) -> list:
+        """
+        Verifies that the provided links visited in private browing session do not appear in the awesome bar.
+        Args:
+            links `list`: A list of links to be verified.
+
+        Returns:
+            list: A list of links that appeared in the awesome bar during private browsing.
+        """
+        initial_window_handle = self.selenium.current_window_handle
+        self.open_private_window()
+        self.switch_to_new_window_or_tab()
+
+        for link in links:
+            self.selenium.get(link)
+
+        self.selenium.switch_to.window(initial_window_handle)
+        return self.awesome_bar(links)
+
+    def verify_links_in_history(self, links: list, new_window: bool = False) -> list:
+        """
+        Verifies that the provided links appear in the history.
+        Args:
+            links `list`: A list of links to be verified.
+
+        Returns:
+            list: A list of links that were not found in the history.
+        """
+        if new_window:
+            self.open_new_window()
+        else:
+            self.open_new_tab()
+        self.switch_to_new_window_or_tab()
+        for link in links:
+            self.selenium.get(link)
+        self.open_panel_menu()
+        self.open_history_menu()
+        urls = []
+        for link in links:
+            if not self.is_present(link):
+                urls.append(link)
+        return urls
+
+    def verify_private_browsing_links_not_in_history(
+        self,
+        links: list,
+    ) -> list:
+        """
+        Verifies that the provided links visited in private browsing session do not appear in the history.
+        Args:
+        links (`list[str]`): List of URLs visited during private browsing session
+        history (`History`): An instance of the History class used to check if a link is present in the history.
+
+        Returns:
+            list: A list of links that were found in the history during the private browsing session.
+        """
+        invalid_links = []
+        initial_window_handle = self.selenium.current_window_handle
+        self.open_private_window()
+        self.switch_to_new_window_or_tab()
+
+        for link in links:
+            self.selenium.get(link)
+
+        self.selenium.switch_to.window(initial_window_handle)
+        self.open_panel_menu()
+        self.open_history_menu()
+        for link in links:
+            if self.is_present(link):
+                invalid_links.append(link)
+        return invalid_links
+
```

Comment:
What's the logic behind these additions? These seem like actions that a user would do. Are you expecting a user to supply a list of URLs for these methods to verify? Why wouldn't they just do that themselves?

Reply:
Yeah, you are right. The actions mirror what a user would do manually, but given that this is a test automation tool, I think its purpose is always to automate manual user actions. Technically, this isn't user functionality - it's testing infrastructure. In this sense, I try to automate certain expected behaviors of Private Window browsing vs Normal Window browsing, history etc.

Reply:
Yeah, the user will have to supply a list of URLs

Reply:
Remember, we are just trying to programmatically represent a way for a user to interact with Firefox. We don't want to assume what combinations they will do, just provide them a way to do them.

Reply:
Okay,  that makes good sense. I will move those methods to tests. So they serve as part of FoxPuppet test suite. Will that be good or I just remove them totally.

Reply:
Desktop test  has similar tests: https://github.com/mozilla/fx-desktop-qa-automation/blob/main/tests/security_and_privacy/test_private_session_awesome_bar_exclusion.py